Add an XRT plugin (#170)
triSYCL/sycl: sycl/doc/GettingStartedXilinxFPGA.md

# You can also try --branch sycl/unified/next for a bleeding edge experience
git clone --branch sycl/unified/master git@github.com:triSYCL/sycl.git llvm
python $SYCL_HOME/llvm/buildbot/configure.py

> python $SYCL_HOME/llvm/buildbot/compile.py --xrt
> # The --xrt is to enable the optional XRT plugin. it is a replacement for the OpenCL plugin.
> # Because the OpencCL implementation in XRT is not maintained anymore

Reviewer: ```suggestion
# Use --xrt is to enable the optional XRT plugin. This is a replacement for the OpenCL plugin
# because XRT offers more control and expressiveness on the hardware
python $SYCL_HOME/llvm/buildbot/compile.py --xrt
```
to be more factual

### Debugging the driver and intermediate steps

During the compilation, a temporary directory is created in which all Vitis inputs, commands, outputs and logs are stored.
By default, this directory is deleted as soon as the compilation ends (even when it fails).

In order to keep it, set the `SYCL_VXX_KEEP_CLUTTER` environment variable to True.

The compiler will output a similar to

```
Temporary clutter in /tmp/EXECNAME-e5ece1pxk5rz43 will not be deleted
```

Informing you of where those files are kept (`/tmp/EXECNAME-e5ece1pxk5rz43` in this case).

### Debugging the SYCL runtime

There is 2 supported backends targeting Xilinx/AMD FPGA the XRT backend and the OpenCL backend.
if you are using the default device selector the xrt backend selected by:
```bash
export SYCL_DEVICE_FILTER=xrt
```
and the OpenCL backend can be selected by:
```bash
export SYCL_DEVICE_FILTER=opencl
```

Testing if it is possible to reproduce a bug on the other backend can give you more information about the bug.

Also the XRT backend support generating reproducer for debugging purposes (or bug reports). To generate a reproducer:

Run the program with SYCL_PI_XRT_REPRODUCER_PATH=path/to/reprod.out.cpp

> create a file named reprod.cpp with:

Reviewer: ```suggestion
Create a file named `reprod.cpp` with:
```

```cpp
#include <xrt/xrt_kernel.h>
#include <xrt.h>
#include <array>
#include <cassert>
#include <cstring>

int main() {

/// insert the code here

/// Edit the code below to validate the data
/// int* a_c = name#.data();
for (int i = 0; i < /*size*/; i++) {
int res = i + i + 1;
int val = a_c[i];
assert(val == res);
}

> printf("PASS\n");

Reviewer: This is not indented?

Author: the code automatically generated by the reproducer is not indented so the surrounding was not indented either.
but we can indent it.

Reviewer: OK.

}
```

> copy the contents of path/to/reprod.out.cpp into reprod.cpp where the comment says so.

Reviewer: ```suggestion
copy the content of `path/to/reprod.out.cpp` into `reprod.cpp` where the comment says so.
```

the find all the TODOs in the file.

> if your sycl code is simple there should be only 1 TODO which should be replaced by the path to the xclbin see the previous section to get the path to the xclbin

Reviewer: ```suggestion
Then find all the TODOs in the file.
If your SYCL code is simple, there should be only 1 TODO which should be replaced by the path to the `xclbin` see the previous section to get the path to the `xclbin`.
```

> also edit the end of the main to adapt or remove the validation of data

Reviewer: ```suggestion
Also edit the end of the `main` to adapt or remove the validation of data
```
